### PR TITLE
Add basic inference rule functionality

### PR DESCRIPTION
**Pull request by serramatutu**
# What does this PR do?

It adds the base classes required for implementing inference rules with minimal effort. Those classes make it easy for custom rules to return signals about columns, associated with confidence scores and reasons. For example, a custom rule could return a signal as follows:
```python
InferenceSignal(
    column=SqlColumn.from_names(db="db", schema="schema", table="table", column="id")
    type=InferenceSignalType.PRIMARY_IDENTIFIER,
    confidence=InferenceSignalConfidence.FOR_SURE,
    reason="The column is named `id`"
)
```

A heuristic that reasons about a collection of signals could then use confidence values to make conclusions about a column, giving well-described reasons to the end user. It could also prevent false positives by detecting conflicting signals (see `InferenceSignal.conflict`).

Apart from that, it also makes it very easy for an `InferenceRunner` to determine which contexts are required by a rule to run without failure by using introspection from type annotations in the `process` method.

Example:
```python
class CtxType1(InferenceContext):
    pass

class CtxType2(InferenceContext):
    pass

class TestRule1(InferenceRule):
    def process(self, ctx1: CtxType1):
        pass

class TestRule2(InferenceRule):
    def process(self, ctx2: CtxType2):
        pass

class TestRule1And2(InferenceRule):
    def process(self, ctx1: CtxType1, ctx2: CtxType2):
        pass

print(TestRule1().required_contexts)  # (CtxType1,)
print(TestRule2().required_contexts)  # (CtxType2,)
print(TestRule1And2().required_contexts)  # (CtxType1, CtxType2)
```

Finally, added a basic rule that produces signals based on a Regex matches to column names.

### CODE REVIEW

File: metricflow/inference/rule/base.py
Comment (serramatutu) [inline metricflow/inference/rule/base.py:134]: I like making the superclass infer the `required_contexts` from class annotations because it makes subclassing very easy and idiomatic, as in the PR description. However, this violates the principle that subclasses must keep the subclass interface. 

I wanna know your opinions on this. Maybe this is just too much magic? I kinda like it :)

Comment (serramatutu) [reply metricflow/inference/rule/base.py:134]: Btw, MyPy complains about this, so you need to add `# type: ignore` to each subclass implementation (which I hate). I haven't found a solution for this, though...

File: metricflow/inference/rule/rules.py
Comment (serramatutu) [inline metricflow/inference/rule/rules.py:32]: Just realized I need to update docs since this does not `match` the full name, but `searches` it.

File: metricflow/test/inference/rule/test_base.py
Comment (serramatutu) [inline metricflow/test/inference/rule/test_base.py:94]: oops

File: metricflow/test/inference/rule/test_rules.py
Comment (serramatutu) [inline metricflow/test/inference/rule/test_rules.py:10]: will also remove this comment

File: metricflow/inference/rule/base.py
Comment (tlento) [inline metricflow/inference/rule/base.py:77]: In thinking about this a bit further I think you want a hierarchy around InferenceSignalType that makes sense as a return.

Fundamentally, you're trying to figure out how to populate field sets in a DataSource. That means you care about, basically, three things: Dimensions, Identifiers, and Measures

So you can have essentially three concrete classes that are something like `InferredColumnType` and each of those can then encapsulate their own hierarchy of inferred attributes, and manage consistency checking internally. They'd implement a shared `InferredColumnType` interface, so you can operate on them interchangeably.

If you ever have something that's shared across those class divisions - like maybe there's a signal that something maybe shouldn't be included in the dimension/identifier/measure set but if the user does want it the inference should remain - you can define another cross-cutting class that essentially extends the functionality across all `InferredColumnType` base classes.

That gives you a much cleaner, more extensible model, full typechecking support, and will naturally dictate the boundaries of what's acceptable in terms of return types for various inference rules. The internals of which class of thing to return might be a little more difficult on the implementer, but everything else will likely be cleaner, and you can smooth out those rough edges as needed.

**Review by tlento — COMMENTED** [review]
So I think I see what you're trying to do here, and it seems good to me. If I'm right, you want:

1. A generic framework that can rely on some contextual data retrieved from somewhere to produce an inference for evaluation by whatever will write out the detected model
2. Some way of allowing people to specify which sorts of contextual data their inference generator requires
3. The ability to share logic across rules of similar type

The main issue with the core implementation laid out here is it does everything in a mostly untyped manner - the core API is effectively untyped, requirements are identified based on runtime lookups of type inspection, and logic is shared by coercing inputs to strings and running regular expressions (in the one concrete example you've provided, anyway).

I've put some design suggestions inline for you before I pack it in for the long weekend. They're not the only way to do this, necessarily, but they're things we do elsewhere in the framework and so they'll look familiar to developers working here, and you can probably poke around and see some examples of sufficiently similar things to get you started.

I'll be happy to answer any questions next week, although given how subtle and difficult some of this stuff can get it might be more efficient for us to jump on a call. I'll leave that up to you to decide, feel free to grab time with me at your convenience.

File: metricflow/inference/rule/defaults.py
Comment (serramatutu) [reply metricflow/inference/rule/defaults.py:20]: I know it is broad, but my rationale is that it's better to match stuff like `userid` and `customerid` at the expense of having a couple of (hopefully rare) false positives like `is_livid` and `grid` than to not match them and have the user type them in everytime. 

It will be very difficult to get a solution that gets everything right everytime, but IMO that's a tolerable compromise (given we make sure the user reviews everything to catch those misses).

**Review by tlento — COMMENTED** [review]
Ok, I'm back here to cover the changes in prep for #163. This seems pretty reasonable for now, given we're in prototype mode, so I will come back and approve once I see how it works in practice over there. :)

**Review by tlento — APPROVED** [review]
This has all really come together nicely. Great stuff! Thanks again for such a well-crafted set of PRs, and for your patience on the back and forth. Love the commit structure and messaging!

**Pull request merged**
